transport/src/main/java/io/netty/channel/DefaultChannelHandlerContext.java

```diff
@@ -130,14 +182,26 @@ public ChannelHandlerContext fireChannelRegistered() {
     }
 
     private void findAndInvokeChannelRegistered() {
-        findContextInbound(MASK_CHANNEL_REGISTERED).invokeChannelRegistered();
+        DefaultChannelHandlerContext context = findContextInbound(MASK_CHANNEL_REGISTERED);
+        if (context.isProcessInboundDirectly()) {
+            context.invokeChannelRegistered();
+        } else {
+            executeInboundReentrance(context, context::invokeChannelRegistered0);
```

Comment:
I think this will only work iff the event loop _guarantees_ that all of the things enqueued with `executeInboundReentrance` are executed before any further socket reads. And I don't think Netty guarantees that : I think there is 'read spin count' which might be a problem and I think you have some 'io vs. tasks ratio' thing which might prevent all `eventloop.execute`d things from running before the next socket read.

If we don't do all the events before the next socket read, there will be potential event reordering , right?

Reply:
@weissi this does not matter as it will continue to offload as long as not all previous offloaded tasks were completed. So there is no reordering issues.

Reply:
@normanmaurer ah right, yes! But it could then lead to pretty bad performance, right?

Reply:
@weissi in theory yes but this would be true for anything that is scheduled on the EventLoop. If you can not keep up with draining the queue you are in trouble anyway if you continue to read from the Channel.